numpy/random/bit_generator.pyx

```diff
@@ -0,0 +1,613 @@
+    return result
+
+
+class ISeedSequence(abc.ABC):
+    """
+    ISeedSequence() is the abstract base class for SeedSequences.
+
+    See Also
+    --------
+    SeedSequence, SeedlessSeedSequence
+    """
+
+    @abc.abstractmethod
+    def generate_state(self, n_words, dtype=np.uint32):
+        """
+        generate_state(n_words, dtype=np.uint32)
+
+        Return the requested number of words for PRNG seeding.
+
+        A BitGenerator should call this method in its constructor with
+        an appropriate `n_words` parameter to properly seed itself.
+
+        Parameters
+        ----------
+        n_words : int
+        dtype : np.uint32 or np.uint64, optional
+            The size of each word. This should only be either `uint32` or
+            `uint64`. Strings (`'uint32'`, `'uint64'`) are fine. Note that
+            requesting `uint64` will draw twice as many bits as `uint32` for
+            the same `n_words`. This is a convenience for `BitGenerator`s that
+            express their states as `uint64` arrays.
+
+        Returns
+        -------
+        state : uint32 or uint64 array, shape=(n_words,)
+        """
+
+
+class ISpawnableSeedSequence(ISeedSequence):
+
+    @abc.abstractmethod
+    def spawn(self, n_children):
+        """
+        spawn(n_children)
+
+        Spawn a number of child `SeedSequence` s by extending the
+        `spawn_key`.
+
+        Parameters
+        ----------
+        n_children : int
+
+        Returns
+        -------
+        seqs : list of `SeedSequence` s
+        """
+
+
+cdef class SeedlessSeedSequence():
+    """
+    SeedlessSeedSequence() is used for BitGenerators with no need for seed state.
+
+    See Also
+    --------
+    SeedSequence, ISeedSequence
+    """
+
+    def generate_state(self, n_words, dtype=np.uint32):
+        raise NotImplementedError('seedless SeedSequences cannot generate state')
+
+    def spawn(self, n_children):
+        return [self] * n_children
+
+
+# We cannot directly subclass a `cdef class` type from an `ABC` in Cython, so
+# we must register it after the fact.
+ISpawnableSeedSequence.register(SeedlessSeedSequence)
+
+
+cdef class SeedSequence():
+    """
+    SeedSequence(entropy=None, program_entropy=None, spawn_key=(), pool_size=4)
+
+    SeedSequence mixes sources of entropy in a reproducable way to set the
+    initial state for independent and very probably non-overlapping
+    BitGenerators.
+
+    Once the SeedSequence is instantiated, you can call the `generate_state`
+    method to get an appropriately sized seed. Calling `spawn(n) <spawn>` will
+    create ``n`` SeedSequences that can be used to seed independent
+    BitGenerators, i.e. for different threads.
+
+    Parameters
+    ----------
+    entropy : {None, int, sequence[int]}, optional
+        The entropy for creating a `SeedSequence`.
+    program_entropy : {None, int, sequence[int]}, optional
+        A second source of entropy, typically per-application
```

Comment:
@rkern how settled is the `program_entropy` idea in your mind? Just to be sure this doesn't sneak in unintentionally as API. And now that I think about it, we could add a `, *,` since this is Py3 only to make all those more advanced arguments keyword only for now.

Reply:
I mean, _I_ like it. I don't know how well my explanations about it went down. It's something that could be removed now and added back later, when there are fewer new features for users to process.

Reply:
And there are ways for me to implement the same concept in my own programs with this argument removed, so nothing but convenience is lost by removing it now.

Reply:
Maybe an example would help me, but it does not matter much, I think I would vote for making it kwarg only. And after that, quite frankly, we have easier flexibility with deprecating it in principle. 

Would this be the use case?:
```python
def algorithm_which_user_can_make_reproducable(..., entropy=None):
    """User can provide entropy or a SeedSequence instance"""
    if not isinstance(entropy, SeedSequence):
        seed_seq = SeedSequence(entropy, program_entropy=0x28374289374)
   else:
       # trust user:
       seed_seq = entropy
```
In which case all users still would end up using similar entropy/seed for this specific function, but this specific function will not use the same ones as the rest of the world in case users put in "bad" entropy?

And yes, for a long time there, I was thinking it was (also?) a solution to the nudging users to use better seeds to allow simple manual seeds to be good entropy by mixing the two (while keeping it easy to try different fixed streams):
```
# rng = np.random.Generator(np.randomPCG64(1))  # produced quirky plot "a.pdf"
rng = np.random.Generator(np.randomPCG64(2))   # used for "b.pdf"
```
although I am not sure that is needed.

One other thing: Should we document some best practices? If users pass SeedSequences to external libraries, it probably makes sense to to be generous about spawning new ones? Since I can control how many SeedSequences I spawened off, but I cannot control how many spawns a library creates?

Reply:
Technically, it would be useful wherever one accepts manually-typed integer seeds. Ideally, in a best-practices world, that's just in your `main()`, and then you thread the same `BitGenerator` through your whole program (spawning or jumping when you need to split things into independent processes).

Things like `check_random_state()`, while good practice, do mean that the _potential_ API surface where integer seeds can be provided is pretty large. People _shouldn't_ pass in an integer seed to each `Estimator` in the same program, though I'm sure someone does, somewhere. It's more forgiveable in a Jupyter notebook context where you can't really be sure about the order your cells execute, so maybe you want a per-cell seed for some cases there. But `program_entropy` isn't that useful in that case either, since you're seeing both side-by-side. If you knew enough to make `program_entropy` big and unique per cell, you would just use a good seed to begin with. `program_entropy` really only useful when the user is entering the main seed in a different way than typing literals into the code (e.g. command-line argument, etc.).

> Should we document some best practices? If users pass SeedSequences to external libraries, it probably makes sense to to be generous about spawning new ones? Since I can control how many SeedSequences I spawened off, but I cannot control how many spawns a library creates?

The only real limitation is 2**32 children per node. That should be enough for everyone.

Reply:
We also probably do not want to do `for i in range(1000): seq = seq.spawn()[0]`, but neither of these are very likely to occur and both could get a warning/error tagged on in principle.

For now I think I would simply prefer to add the `, *,`. Another, probably not plausible API would be to make it `(entropy, (small_int, spawn_key), (small_int, spawn_key))` allowing `.spawn()` as well as `.seeded()`. That would give me a way to create the same random numbers twice without saving the state.

Reply:
So sorry, about the discussion. If you are happy with it, I would prefer adding `,* ,`, and we can leave it at that.

Reply:
I think removing `program_entropy` for now would be fine. And yes, it should be `entropy=None, *, ...` whatever follows.

Reply:
removed program_entropy, added kw-only `*`